test(compiler-cli): error when running tests on non-posix systems

crisbeto opened the pull request:
We weren't resolving a path correctly which resulted in an error on Windows. For reference, here's the error. Note the extra slash before `C:`:

```
Error: ENOENT: no such file or directory, scandir '/C:/bazel_output_root/yxvwd24o/external/npm/node_modules/typescript'
    at Object.readdirSync (fs.js:854:3)
```

JoostK commented on packages/compiler-cli/ngcc/test/integration/util.ts:
```diff
@@ -224,14 +224,15 @@ export function compileIntoApf(
       fs.resolve(`/node_modules/${pkgName}/package.json`), JSON.stringify(pkgJson, null, 2));
 }
 
+setFileSystem(new NodeJSFileSystem());
```
This call is suspect, as `loadStandardTestFiles` should be able to execute without a globally set filesystem. I checked locally and it turns out the call to `togglePathCase` depends on it:

https://github.com/angular/angular/blob/2c4a98a28594afc16a481ff4fc56cb37ce1a04b0/packages/compiler-cli/src/ngtsc/file_system/src/node_js_file_system.ts#L76

https://github.com/angular/angular/blob/2c4a98a28594afc16a481ff4fc56cb37ce1a04b0/packages/compiler-cli/src/ngtsc/file_system/src/node_js_file_system.ts#L132-L135

Removing the `absoluteFrom` from `togglePathCase` and changing its return type to `string` should be sufficient to avoid this setter here.

JoostK reviewed (changes requested):
Sorry for breaking your workflow and thanks for the fix! I was curious how this worked before, as this area has always interacted with the native filesystem and worked correctly. The POSIX variant we use in these tests are only for in-memory interactions, so it was unexpected to me that the used FS had any impact.

Upon tracing back the changes I did in #38909, I noticed the following is incorrect:

https://github.com/angular/angular/blob/4820b690c12825b0ee6d1b194bf741e55ddb4961/packages/compiler-cli/test/helpers/src/mock_file_loading.ts#L80

The `tmpFs.resolve(path)` should just be `path`, as `path` is a native path and should stay that way without `tmpFs.resolve` incorrectly prefixing the native path with a slash.

```diff
- loadTestDirectory(tmpFs, tmpFs.resolve(path), tmpFs.resolve('/')); 
+ loadTestDirectory(tmpFs, path, tmpFs.resolve('/')); 
```

crisbeto commented:
Thanks for the feedback! My changes seemed a little fishy, but I wasn't familiar enough with the mock file systems to know the best place to fix it. I've applied both of your suggestions.

petebacondarwin commented on packages/compiler-cli/test/helpers/src/mock_file_loading.ts:
```diff
@@ -77,7 +77,7 @@ export function loadFakeCore(fs: FileSystem, basePath: string = '/') {
 function loadFolder(path: string): Folder {
   const tmpFs = new MockFileSystemPosix(true);
-  loadTestDirectory(tmpFs, tmpFs.resolve(path), tmpFs.resolve('/'));
+  loadTestDirectory(tmpFs, path, tmpFs.resolve('/'));
```
Nice spot @JoostK - I should have noticed this when reviewing your PR - I have fallen foul of making this change by mistake myself, before. Can we add a comment above this line to ensure that it doesn't happen again?

crisbeto replied:
Added.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/file_system/src/node_js_file_system.ts:
```diff
@@ -129,7 +128,6 @@ export class NodeJSFileSystem implements FileSystem {
-  return absoluteFrom(
-      str.replace(/\w/g, ch => ch.toUpperCase() === ch ? ch.toLowerCase() : ch.toUpperCase()));
+function togglePathCase(str: string): string {
+  return str.replace(/\w/g, ch => ch.toUpperCase() === ch ? ch.toLowerCase() : ch.toUpperCase());
```
@JoostK can you explain why the `absoluteFrom` here breaks things? This should only be called when the FS system is `NodeJSFileSystem` so using `absoluteFrom` here should work...

JoostK replied:
It used to break in the original commit in this PR, as the `MockFileSystemNative` was used which delegates to `NodeJSFileSystem`.

If we do want to change this still, it's probably best to do so in a separate commit. WDYT, @petebacondarwin?

petebacondarwin replied:
The problem, I believe, is that you cannot guarantee that the `str` passed to `togglePathCase()` is actually normalized according to the `MockFileSystemNative`. If this PR passes tests without this change, I would avoid it.

JoostK replied:
Well, the result of `togglePathCase` is only used for native `fs.existsSync` and we pass in `__filename`, so I figured we don't risk it being non-normalized. But I agree that we don't need to have this change in this PR.

petebacondarwin replied:
On Windows `__filename` would look like: `C:\Users\Admin\dev\angular\angular\test.js` but I believe that even on the "native" file-system we normalize the backslashes to forward slashes: `C:/Users/Admin/dev/angular/angular/test.js`.

petebacondarwin replied:
Removing this change doesn't break `ngcc/integration` tests. So I would rather drop it from this PR.

JoostK replied:
Correct; the original changes to switch to `MockFileSystemNative` were undone, so this change is no longer necessary to fix the Windows issue.

crisbeto replied:
I was under the impression that it was something we wanted to do regardless of the Windows issue. I've reverted it.

crisbeto committed:
test(compiler-cli): error when running tests on non-posix systems

We weren't resolving a path correctly which resulted in an error on Windows.
For reference, here's the error. Note the extra slash before `C:`:

```
Error: ENOENT: no such file or directory, scandir '/C:/bazel_output_root/yxvwd24o/external/npm/node_modules/typescript'
    at Object.readdirSync (fs.js:854:3)
```
